[Form][TwigBridge] Render an id on the form element so form_attr references resolve

ker0x opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #42075
| License       | MIT

`BaseType` puts the root form's id in `vars['id']`, and the themes render it on the element wrapping the fields, never on the `<form>` tag. The `form_attr` option points children at that same id. HTML requires `form="X"` to reference a `<form id="X">`, so the reference has always pointed at a `<div>` and the option has never worked.

A new `form_id` view variable carries the id of the `<form>` element. It holds whatever that element will actually carry:

* the `attr.id` set on the root form, when the application chose one;
* the string given to `form_attr`, when that option is a string, on the root form or on any descendant, since that is the identifier the application already picked for this very purpose;
* otherwise `form_` followed by `vars['id']`, prefixed so that the id of the wrapping element does not move.

`FormType::finishView()` computes it. It runs on the root form after every child view exists, so it is the first point where the `form_attr` option of every descendant is known, and it is where the children that use the option are pointed at the value. The tag and the references are therefore written in one place and cannot drift apart.

The variable stays `null` when no descendant uses `form_attr`, and the theme then renders nothing. Forms that do not use the option, which is nearly all of them, render exactly as before.

`form_div_layout.html.twig` fills `attr.id` with the variable when it is set and the application has not chosen an id itself. Set `attr.id` on the root form to choose the id, or `attr: {id: false}` to render none, as for any attribute.

Before:

```html
<form name="my_form" method="post" action="/submit">
  <div id="my_form">
    <input id="my_form_first" name="my_form[first]" form="my_form" />
```

After:

```html
<form name="my_form" method="post" action="/submit" id="form_my_form">
  <div id="my_form">
    <input id="my_form_first" name="my_form[first]" form="form_my_form" />
```

A form with no `form_attr` anywhere is untouched:

```html
<form name="my_form" method="post" action="/submit">
```

All themes inherit `form_start` from `form_div_layout`, so all of them gain the id and none needed its own change. A custom theme that overrides `form_start` renders no id until it is updated.

The `form_attr` string form keeps its meaning: it still sets `vars['id']` for the root when it is set there, descendant ids and `label for=` are unaffected, and children keep referencing the string itself. What it gains is that the `<form>` element now renders that id, so the reference resolves without the application having to repeat the identifier in `attr.id`.

Documentation should cover: the `form_id` view variable, how it is derived and when it is set, that `form_attr` now resolves, and the two ways to control the id (`attr.id` for a custom one, `attr: {id: false}` for none).

ker0x committed:
[Form] Add id attribute to <form> element and use row_attr for form tag attributes

ker0x commented on src/Symfony/Bridge/Twig/Resources/views/Form/form_div_layout.html.twig:
```diff
@@ -18,7 +18,7 @@
 
 {%- block form_widget_compound -%}
-    <div {{ block('widget_container_attributes') }}>
+    <div{% if not form.methodRendered %} {{ block('widget_container_attributes') }}{% endif %}>
```
After talking with @xabbuh, it seems that `form.methodRendered` is used to ensure that the _method field is not rendered twice, which is not what I am trying to do here.

Is there any way to know if a `form` is the `<form>` element?

maxbeckers replied:
Perhaps you could try it with `form.parent is null`? Did not try, just an idea.

ker0x replied:
It was my first attempt, using the rootform check. But if you do something like that:

```php
$form = $this->createForm(
    CollectionType::class,
    [
        'foo@example.com',
        'bar@example.com',
    ],
    [
        'entry_type' => EmailType::class,
        'entry_options' => [
            'attr' => ['class' => 'datetime-box'],
        ],
        'attr' => [
            'class' => 'foo',
            'data-foo' => 'attr',
        ],
        'row_attr' => [
            'class' => 'bar',
            'data-foo' => 'row-attr',
        ],
    ]);
```
And only calling `{{ form_widget(form) }}` in your template without `form_start`, then the rootform is the collection and it's ID and attribures will not be displayed.

xabbuh replied:
Making any change here looks like a BC break to me anyway as developers do probably rely on the id always being rendered here. IMO we would need a different id for the wrapping `<form>` element.

ker0x replied:
How do you think this new form ID should be defined? Through a new option on the `FormType` or directly in the `form_start` block by concatenating for example `form_{{ id }}`?

xabbuh replied:
I think we do not need an option for this. But we could compute it when building the `FormView`.

ker0x commented:
I just updated the PR with the following changes:

- A new calculated variable `form_id` is available in the view. It is the concatenation of `form_` and the `ID` of the form.
- The `<form>` element uses the `row_attr` option instead of `attr` to display attributes. The `row_attr` option was never used on the root form and this avoids duplicating attributes.
- The `form_attr` option now only accepts a boolean value. 
    - If set on the root form, each child will have a `form` attribute with the `form_id` as value.
    - If set on a child, it will have a `form` attribute with the `form_id` as value.

nicolas-grekas commented:
Looks like a rebase is needed. Still up to do it?

ker0x commented:
@nicolas-grekas PR has been rebase, however unit tests are failing with high and low deps and I don't know why!

xabbuh commented:
looks like you need to update the `composer.json` file of the Twig bridge to make sure that the Form component is not older than 6.3 (in the `require-dev` and `conflict` sections)

nicolas-grekas commented:
I pushed the fix to the composer.json.
Can you please update the description of the PR? I feel like it's a bit out of sync.

HeahDude reviewed (approved):
Since this issue has been there for a long time already, maybe this should target 6.4 to prevent side effects in the LTS.

ker0x commented:
@nicolas-grekas description update 👍 

@HeahDude I agree to target `6.4` instead of `5.4`

nicolas-grekas commented:
> target 6.4 instead of 5.4

Would be better I agree. Can you update the target of this PR and rebase on 6.4 please?

xabbuh reviewed (changes requested):
you will need to update the version constraints similarly to what Nicolas did in https://github.com/symfony/symfony/commit/3f3d7634f26767941f5a1525b7a4b912845b419c#diff-33dca46fccec7efe765aa6cf249fddcdde0d15fa2d01adc052a5e1da01b423c5

xabbuh commented on src/Symfony/Bridge/Twig/composer.json:
```diff
@@ -28,7 +28,7 @@
         "symfony/dependency-injection": "^5.4|^6.0|^7.0",
         "symfony/finder": "^5.4|^6.0|^7.0",
-        "symfony/form": "^6.3|^7.0",
+        "symfony/form": "~6.3.2|^7.0",
```
```suggestion
        "symfony/form": "^6.3.2|^7.0",
```

xabbuh commented on src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_horizontal_layout.html.twig:
```diff
@@ -1,7 +1,7 @@
 
 {% block form_start -%}
-    {% set attr = attr|merge({class: (attr.class|default('') ~ ' form-horizontal')|trim}) %}
+    {% set row_attr = row_attr|merge({class: (row_attr.class|default('') ~ ' form-horizontal')|trim}) %}
```
If we have to make this change here, it seems to me that we need to adapt for custom form types that do not provide the `row_attr` variable (and that we need to fall back to `attr` then).

nicolas-grekas reviewed (approved):
@xabbuh can you please have another look?
I just rebased on top of 8.1 + addressed a few things (esp BC considerations.)

Nayte91 commented:
I would add that if this PR is merged, then we would need to add the "form" attribute on Symfony's generated submit buttons to wire those buttons on forms. I think about the Multisteps form's previous/next/finish, for example.

ker0x committed:
[Form][TwigBridge] Render an id on the form element so form_attr references resolve

The root form id is rendered on the element wrapping the fields, never on the
<form> tag, while the form_attr option points children at that same id. HTML
requires form="X" to reference a <form id="X">, so the reference has always
pointed at a div and the option never worked.

A new form_id view variable carries the id of the <form> element: the one set
through attr.id when the application chose it, the string given to form_attr
when that option is a string, wherever it sits in the form tree, and otherwise
form_ followed by the root form's id so that no existing id moves. FormType
fills it in finishView(), the first point where the option of every descendant
is known, and points the children that use form_attr at that same value. The
default theme fills attr.id with it.

xabbuh commented on src/Symfony/Component/Form/Extension/Core/Type/BaseType.php:
```diff
@@ -66,10 +66,14 @@ public function buildView(FormView $view, FormInterface $form, array $options):
+
+                // must match the id the theme renders on the <form> element, not the one of the
+                // wrapping element, otherwise the reference points at nothing
+                $options['attr']['form'] = self::formId($rootId);
```
I would not add the `form_` prefix if `form_attr` is a string. It's likely that a string was used before to work around the fact that we never rendered an `id` attribute on the `form` element and that people had their own logic or form theme.

xabbuh commented on src/Symfony/Component/Form/Extension/Core/Type/BaseType.php:
```diff
@@ -147,4 +156,23 @@ public function configureOptions(OptionsResolver $resolver): void
+            return $options['attr']['id'];
+        }
+
+        return \is_string($options['form_attr']) ? $id : 'form_'.$id;
```
I am not sure this works as expected. `$options` is the options of the root form, but we need to account for `form_attr` used on any of its children.

nicolas-grekas replied:
Fixed: `formId()` is gone, `FormType::finishView()` now derives the id from the whole form tree.

xabbuh commented on src/Symfony/Bridge/Twig/Tests/Extension/AbstractLayoutTestCase.php:
```diff
@@ -2381,6 +2381,65 @@ public function testStartTag()
+        $form = $this->factory->createNamedBuilder('registration', 'Symfony\\Component\\Form\\Extension\\Core\\Type\\FormType', null, [
+            'method' => 'get',
+            'form_attr' => 'custom-identifier',
+        ])->add('username', 'Symfony\\Component\\Form\\Extension\\Core\\Type\\TextType')->getForm();
```
```suggestion
        ])->add('username', 'Symfony\\Component\\Form\\Extension\\Core\\Type\\TextType', ['form_attr' => 'custom-identifier'])->getForm();
```

nicolas-grekas replied:
Applied.

xabbuh commented on src/Symfony/Bridge/Twig/Tests/Extension/AbstractLayoutTestCase.php:
```diff
@@ -2381,6 +2381,65 @@ public function testStartTag()
+
+        $form = $this->factory->createNamedBuilder('registration', 'Symfony\\Component\\Form\\Extension\\Core\\Type\\FormType', null, [
+            'method' => 'get',
+            'form_attr' => 'custom-identifier',
```
```suggestion
```
